DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/svd/src/svd.hpp

```diff
@@ -0,0 +1,273 @@
+                                  EigenVectorsPipe, RankDeficientFlagPipe>());
+
+  // read input matrix from DDR (again)
+  auto ddr_write_event_post = q.submit([&](sycl::handler &h) {
+    h.single_task<IDInputMatrixFromDDRToLocalMemPostProcess>(
+        [=]() [[intel::kernel_args_restrict]] {
+          MatrixReadFromDDRToPipeByBlocks<
+              T, cols, rows, kNumElementsPerDDRBurst, InputMatrixPipe2>(
+              input_matrix_device, 1, repetitions);
+        });
+  });
+
+  // Post process. Using eigen values and eigen vector to produce U, S, V matrix
+  q.single_task<IDUSVFromEigens>(
+      fpga_svd::PostProcess<T, false, rows, cols, kNumElementsPerDDRBurst,
+                            InputMatrixPipe2, EigenValuesPipe, EigenVectorsPipe,
+                            UTempMatrixPipe, SMatrixPipe, VMatrixPipe>{});
+
+  // Orthogonalizing the U matrix by running it through QRD
+  q.single_task<IDOrthogonalizingU>(
+      fpga_linalg::StreamingQRD<T, false, rows, rows, k_raw_latency,
+                                kNumElementsPerDDRBurst, UTempMatrixPipe,
+                                UMatrixPipe, RMatrixPipe>());
+
+  // Write the rank deficient flag from local memory to FPGA DDR
+  auto rank_deficient_flag_event =
+      q.single_task<IDRankDeficientFlagFromLocalMemToDDR>(
+          [=]() [[intel::kernel_args_restrict]] {
+            VectorReadPipeToDDR<ac_int<1, false>, 1, RankDeficientFlagPipe>(
+                rank_deficient_flag_device, 1, repetitions);
+          });
+
+  // collect R matrix from the orthogonalizing kernel
+  q.single_task<IDRMatrixFromLocalMemToDDR>(
+      [=]() [[intel::kernel_args_restrict]] {
+        for (int rep = 0; rep < repetitions; rep++) {
+          for (int r_idx = 0; r_idx < kRMatrixSize; r_idx++) {
+            r_device[r_idx] = RMatrixPipe::read();
+          }  // end of r_idx
+        }    // end of rep
+      });
+
+  // collecting U matrix from pipe into DDR
+  auto u_matrix_event = q.single_task<IDUMatrixFromLocalMemToDDR>(
+      [=]() [[intel::kernel_args_restrict]] {
+        MatrixReadPipeToDDR<T, rows, rows, kNumElementsPerDDRBurst,
+                            UMatrixPipe>(u_matrix_device, 1, repetitions);
+      });
+
+  // collecting s matrix from pipe into DDR
+  auto s_matrix_event = q.single_task<IDSMatrixFromLocalMemToDDR>(
+      [=]() [[intel::kernel_args_restrict]] {
+        MatrixReadPipeToDDR<T, rows, cols, kNumElementsPerDDRBurst,
+                            SMatrixPipe>(s_matrix_device, 1, repetitions);
+      });
+
+  // collecting V matrix from pipe into DDR
+  auto v_matrix_event = q.single_task<IDVMatrixFromLocalMemToDDR>(
+      [=]() [[intel::kernel_args_restrict]] {
+        MatrixReadPipeToDDR<T, cols, cols, kNumElementsPerDDRBurst,
+                            VMatrixPipe>(v_matrix_device, 1, repetitions);
+      });
+
+  // Wait for all 3 outputs are collected
+  u_matrix_event.wait();
+  s_matrix_event.wait();
+  v_matrix_event.wait();
```

Comment:
https://hsdes.intel.com/appstore/article/#/14019730754

there's apparently an issue with hangs sometimes showing up if you don't `wait()` on all kernels. I'm working on reproducing this.

Reply:
Interesting. Since all 4 computation kernels have while(1) in it to continue process data, we can't wait on them here.